design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,613 @@
+
+Authentication factors are treated as a bloc in the evaluation of alternatives. That is, if in
+a flow there's "Identity provider redirector", "authentication factor 1", "authentication
+factor 2", entering the authentication factor 1 flow will automatically cause "authentication
+factor 2" to be executed after.
+
+To make things perhaps a little more clear, the current default "Browser" authentication
+would become for example:
+
+```
+Auth type
+------------------------------------------------------------
+Identity Provider Redirector
+Authentication factor (1)
+  |- Cookie
+  |- Kerberos
+  |- Username Password Form
+Authentication factor (2) [OPTIONAL]
+  |- Cookie
+  |- Kerberos
+  |- OTP Form
+```
+And if we wanted to have an mandatory 2nd factor, with either OTP or a FIDO2 configured:
+
+```
+Auth type
+------------------------------------------------------------
+Identity Provider Redirector
+Authentication factor (1)
+  |- Cookie
+  |- Kerberos
+  |- Username Password Form
+Authentication factor (2)
+  |- Cookie
+  |- Kerberos
+  |- OTP Form
+  |- FIDO-2
+```
+Potentially we could also introduce another type of authentication execution, we could call it "Authenticated on", to simplify the authenticators that bypass the authentication factors. We could then have:
+
+```
+Auth type
+------------------------------------------------------------
+Identity Provider Redirector
+Authenticated on
+  |- Cookie
+  |- Kerberos
+Authentication factor (1)
+  |- Username Password Form
+Authentication factor (2)
+  |- OTP Form
+  |- FIDO-2
+```
+
+#### Authentication categories
+
+Authentication categories allow the link between the actual execution within the flow and the
+credentials of a certain type.
+
+For example, currently there is the OTPFormAuthenticator for the browser flow and ValidateOTP for the direct grant flow.
+However, both these authentication executions will accept input from the same OTP credentials. This means that schematically, we would have:
+```
+-----------------------         ----------------------
+| OTP Credential 1    |---------|                    |    ------------------------
+-----------------------         |                    |----| OTPFormAuthenticator |
+-----------------------         | OTP Authentication |    ------------------------
+| OTP Credential 2    |---------|      Category      |    ---------------
+-----------------------         |                    |----| ValidateOTP |
+-----------------------         |                    |    ---------------
+| OTP Credential 3    |---------|                    |
+-----------------------         ----------------------
+```
+
+This schema serves as a base for all credentials and allows them to be
+used for the appropriate authentication executions. This link allows Keycloak to present
+alternative credentials for an execution during login.
+
+To simplify selection in the GUI, it would also be useful to do some refactoring to ensure
```

Comment:
+1 for improving this. Currently when you're adding the Authentication Execution, there is no way for "restrict" authenticator implementations based on the context (flow type). So for example ValidateOTP Authenticator is visible even if you're creating "browser" flow, but it shouldn't be displayed as it is applicable just for the direct grant flow (Browser flow has OTPFormAuthenticator).

Perhaps we should add more flow types (Currently there is just "generic" and "client" as you pointed) and each applicable Authenticator (or AuthenticatorFactory?) can have method like "getApplicableFlowTypes", which will return list of flow types for which the Authenticator implementation is valid. Some implementations are valid for more flow types (EG. OTPFormAuthenticator is applicable for "browser" or "post-broker-login" flows and maybe even in some others).

I am not sure if this can be done as part of your work or rather as separate thing?

Reply:
+1 I was thinking we should introduce flow types where an authenticator can advertise which flows it is compatible with. We can introduce it as a new optional method that returns all flow types, and update the built-in authenticators to only support those that they should be used with. I'd say that is general enhancements and doesn't need to be covered by this design proposal though.

Reply:
I also feel that this shouldn't be part of the design proposal. Should I leave the suggesting of the improvement in the text, but explicitly say that it is out of scope for this design document, or just remove the text?